### PR TITLE
Add demo for concurrent web clients

### LINKED ISSUE #5: Add demo
The `main.py` script comes close to this, but would be great to have more formal, basic, step-by-step documentation on how this is intended to be used (tests should mirror if not outright copy or use the same setup to ensure it never breaks)

### PR DESCRIPTION
**Pull request by garrettmflynn**
fix #5 

Just want to confirm this meets our requirements for the demo before switching to a non-draft PR.

### CODE REVIEW

File: src/tqdm_publisher/publisher.py
Comment (CodyCBakerPhD) [inline src/tqdm_publisher/publisher.py]: Is there a reason you prefer the hexidecimal form (without dashes)? Just curious

File: README.md
Comment (CodyCBakerPhD) [inline README.md]: Could you expand this details of how someone would run the demo? Do you they just call Python on the `demo/main.py` file and open the HTML file in any web viewer?

File: demo/main.py
Comment (CodyCBakerPhD) [inline demo/main.py:1]: I'd suggest trying to come up with more specific qualifiers with respect to the `client.html` setup; `python_backend.py` perhaps? `main` might be a bit too generic in this context

File: src/tqdm_publisher/publisher.py
Comment (garrettmflynn) [reply src/tqdm_publisher/publisher.py]: It doesn't really matter to me, aside from that it's a string rather than a UUID type. Do you have a preference?

File: README.md
Comment (garrettmflynn) [reply README.md]: Yeah you got it. I've updated the docs, but how would you want to specify in the demo's `requirements.txt` that it depends on an editable install of the repo's module?

File: demo/main.py
Comment (garrettmflynn) [reply demo/main.py:1]: Is `client.html` and `server.py` specific enough? Or would you prefer some additional qualifiers?

File: src/tqdm_publisher/publisher.py
Comment (CodyCBakerPhD) [reply src/tqdm_publisher/publisher.py]: Oh, I forgot that we usually wrap `uuid4` in a string to get around that (`str(uuid4())`)

I'm personally more used to hashes / IDs having separators in them for whatever reason, but I might be biased. What have you seen more often?

File: README.md
Comment (CodyCBakerPhD) [reply README.md]: Why does it? So that they have a Python file to call?

One of the next things I was going to request (and this ought to be easy with the `pyproject.toml` build strategy) is to expose a CLI entrypoint to run the demo (which ought to then work no matter what the installation pathway)

File: demo/main.py
Comment (CodyCBakerPhD) [reply demo/main.py:1]: That makes sense to me!

Comment (CodyCBakerPhD) [reply demo/main.py:1]: Although... Is a simple script really enough to count as a 'server'? Since it's not a Flask app

Comment (garrettmflynn) [reply demo/main.py:1]: It's a WebSocket server, which is an HTTP server with an additional layer for bidirectional communication.

Comment (CodyCBakerPhD) [reply demo/main.py:1]: Ah, gotcha.

File: README.md
Comment (garrettmflynn) [reply README.md]: Just so it can use `tqdm_publisher`. Maybe there's a better way to reference the module locally?

Do you have any docs for the CLI entrypoint using `pyproject.toml`? Otherwise I'll dig and see if I can find anything that references that.

File: src/tqdm_publisher/publisher.py
Comment (garrettmflynn) [reply src/tqdm_publisher/publisher.py]: I associate separators with IDs that are truly unique for production applications. Otherwise, I just assume it's merely random

File: README.md
Comment (CodyCBakerPhD) [reply README.md]: https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#creating-executable-scripts

> Just so it can use tqdm_publisher. Maybe there's a better way to reference the module locally?

There are, but you wouldn't want them in a demo script. If they've done `pip install tqdm_publisher`, it should work fine (to run the demo, they do need something installed somehow)

File: src/tqdm_publisher/publisher.py
Comment (garrettmflynn) [reply src/tqdm_publisher/publisher.py]: I've updated to use `str(uuid4())` so we can maintain the original formatting.

File: README.md
Comment (garrettmflynn) [reply README.md]: Got it. Just added `tqdm_publisher` to the requirements and switched the base branch to add the executable script.

File: demo/requirements.txt
Comment (CodyCBakerPhD) [inline demo/requirements.txt:1]: Using an additional requirements file negates the push towards a common `pyproject.toml`; I'd wait until #11 is merged and then utilize the 'extra requirements' functionality so that the install pattern in the instructions is

```
pip install tqdm_publisher[demo]
```

mostly to get that extra websockets dependency in to enable the demo even though it's not required for the minimal package usage

File: README.md
Comment (CodyCBakerPhD) [inline README.md]: The point of exposing the entrypoint in the TOML is so that the key in the TOML should become registered as a global command

That is, any Python/conda environment with this package installed, you should be able to run

```
tqdm_publisher-demo-server
```

from the terminal and it should do the same thing as calling python on the file (main difference being, you don't need to know where that file actually lives)

File: README.md
Comment (CodyCBakerPhD) [inline README.md]: Likewise, might want to make a simple CLI exposure to make this call without having to know where the HTML file is

File: pyproject.toml
Comment (CodyCBakerPhD) [inline pyproject.toml]: Since by above, the key should be the thing exposed as a terminal command, the mixture of underscores and dashes makes it look a bit awkward

I'd usually expect a pattern such as `tqdm_publisher run_demo_server` or `tqdm_publisher --run_demo_server`, which would involve altering the `__name__ == "main"` part to be modified to accept input arguments. To avoid that, I wonder if `tqdm_publisher.run_demo_server` is allowed?

File: demo/requirements.txt
Comment (garrettmflynn) [reply demo/requirements.txt:1]: Ah I see, so this removes the need for a requirements.txt file entirely.

Comment (CodyCBakerPhD) [reply demo/requirements.txt:1]: Yep; the goal of the `pyproject.toml` movement is to have 'one configuration file to rule them all'

File: pyproject.toml
Comment (garrettmflynn) [reply pyproject.toml]: Updated to call a `cli.py` script that parses all of these commands. The documentation has been updated accordingly.

File: README.md
Comment (CodyCBakerPhD) [inline README.md]:
```suggestion
tqdm_publisher_demo
```

Thanks, this is looking a lot better now!

I'd still recommend attaching "demo" to the CLI verbage to make it clear this is not in and of itself a CLI package (like, running `tqdm_publisher` on terminal won't magically make a users existing tqdm calls route through our new class)

Comment (CodyCBakerPhD) [reply README.md]: Oops, you beat me to it lol

Comment (garrettmflynn) [reply README.md]: Knew it was coming ;)

**Comment by CodyCBakerPhD** [comment]
I'll try this out on my end after first package publication; looking good to proceed now though

I do wonder if hatchling will be smart enough to bundle the `cli.py` though

**Review by CodyCBakerPhD — APPROVED** [review]

**Pull request merged**
